File: android/app/src/main/java/com/mathgate/MaDateFormat.kt
package com.mathgate

import java.net.URLEncoder
import java.time.LocalDateTime
import java.util.Locale

/**
 * Math Academy's /api/previous-tasks/{date} expects the JavaScript Date.toString() shape, e.g.
 * "Thu Sep 03 2026 11:30:00 GMT-0800 (Pacific Standard Time)", URL-encoded the way
 * encodeURIComponent does it. The PST label is fixed regardless of the caller's zone
 * (this mirrors the community mathacademy-stats extension, which is known to work).
 */
object MaDateFormat {
    private val DAYS = arrayOf("Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat")
    private val MONTHS = arrayOf("Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec")

    fun format(dt: LocalDateTime): String {
        val dow = DAYS[dt.dayOfWeek.value % 7] // java DayOfWeek: MONDAY=1 .. SUNDAY=7
        return String.format(
            Locale.US,
            "%s %s %02d %d %02d:%02d:%02d GMT-0800 (Pacific Standard Time)",
            dow, MONTHS[dt.monthValue - 1], dt.dayOfMonth, dt.year, dt.hour, dt.minute, dt.second,
        )
    }

    fun pathSegment(dt: LocalDateTime): String = encodeUriComponent(format(dt))

    /** java.net.URLEncoder is form-encoding; undo the differences from JS encodeURIComponent. */
    fun encodeUriComponent(s: String): String =
        URLEncoder.encode(s, "UTF-8")
            .replace("+", "%20")
            .replace("%21", "!")
            .replace("%27", "'")
            .replace("%28", "(")
            .replace("%29", ")")
            .replace("%7E", "~")
}

File: ios/MathGateKit/Sources/MathGateKit/MaDateFormat.swift
import Foundation

/// Math Academy's `/api/previous-tasks/{date}` expects the JavaScript `Date.toString()` shape,
/// e.g. `Thu Sep 03 2026 11:30:00 GMT-0800 (Pacific Standard Time)`, encoded the way
/// `encodeURIComponent` does it. The PST label is fixed regardless of the caller's zone; this
/// mirrors the community mathacademy-stats extension, which is known to work.
public enum MaDateFormat {
    private static let days = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"]
    private static let months = [
        "Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec",
    ]

    /// `date` is read in `calendar`'s time zone, so the wall-clock components match the user's day.
    public static func format(_ date: Date, calendar: Calendar = .autoupdatingCurrent) -> String {
        let c = calendar.dateComponents(
            [.year, .month, .day, .hour, .minute, .second, .weekday], from: date
        )
        // Calendar.weekday is 1 = Sunday, matching the array above.
        let dow = days[(c.weekday ?? 1) - 1]
        let month = months[(c.month ?? 1) - 1]
        return String(
            format: "%@ %@ %02d %d %02d:%02d:%02d GMT-0800 (Pacific Standard Time)",
            dow, month, c.day ?? 1, c.year ?? 1970, c.hour ?? 0, c.minute ?? 0, c.second ?? 0
        )
    }

    public static func pathSegment(_ date: Date, calendar: Calendar = .autoupdatingCurrent) -> String {
        encodeURIComponent(format(date, calendar: calendar))
    }

    /// Matches JavaScript's `encodeURIComponent`: everything escaped except these unreserved marks.
    public static func encodeURIComponent(_ s: String) -> String {
        var allowed = CharacterSet.alphanumerics
        allowed.insert(charactersIn: "-_.!~*'()")
        return s.addingPercentEncoding(withAllowedCharacters: allowed) ?? s
    }
}
